File: sort.py
def selectionSort(nums):
    n = len(nums)
    for i in range(n):
        low = i
        for j in range(i + 1, n):
            if nums[j] < nums[low]:
                low = j
        temp = nums[low]
        nums[low] = nums[i]
        nums[i] = temp
# Time Complexity: O(n^2)
# Space Complexity: O(1)

def insertionSort(nums):
    n = len(nums)
    for i in range(1, n):
        j = i
        while j > 0 and nums[j] < nums[j-1]:
            temp = nums[j]
            nums[j] = nums[j-1]
            nums[j-1] = temp
            j -= 1
# Time Complexity: O(n^2) best case is O(n)
# Space Complexity: O(1)

def bubbleSort(nums):
    n = len(nums)
    for i in range(n):
        swapped = False
        for j in range(1, n - i): 
            if nums[j] < nums[j-1]:
                temp = nums[j]
                nums[j] = nums[j-1]
                nums[j-1] = temp
                swapped = True
        if not swapped: break
# Same complexity as insertion
# takes num out unsorted while selection takes num into sorted

arr = [7,2,5,8,9,5,1,3,4]
arr2 = [7,2,5,8,9,5,1,3,4]
arr3 = [7,2,5,8,9,5,1,3,4]
selectionSort(arr)
insertionSort(arr2)
bubbleSort(arr3)
print(arr)
print(arr2)
print(arr3)

File: Java/sort.java
package Java;

import java.util.Arrays;

public class sort{
    public static void main(String[] args){
        int[] nums = {7,3,1,2,7,98,4,2,4,5,7,3,24,5};
        bubbleSort(nums);
        System.out.println(Arrays.toString(nums));
    }
    public static void selectionSort(int[] nums){
        int n = nums.length;
        for(int i = 0; i < n; i++){
            int low = i;
            for(int j = i+1; j < n; j++){
                if(nums[j] < nums[low]){
                    low = j;
                }
            }
            int temp = nums[i];
            nums[i] = nums[low];
            nums[low] = temp;
        }
    }
    public static void insertionSort(int[] nums){
        int n = nums.length;
        for(int i = 1; i < n; i++){
            int j = i;
            while(j > 0 && nums[j] < nums[j-1]){
                int temp = nums[j];
                nums[j] = nums[j-1];
                nums[j-1] = temp;
                j -= 1;
            }
        }
    }
    public static void bubbleSort(int[] nums){
        for(int i = 0; i < nums.length; i++){
            boolean swapped = false;
            for(int j = 1; j < nums.length - i; j++){
                if(nums[j] < nums[j-1]){
                    int temp = nums[j];
                    nums[j] = nums[j-1];
                    nums[j-1] = temp;
                    swapped = true;
                }
            }
            if(swapped == false){
                break;
            }
        }
    }
}